src/crimson/osd/main.cc

```diff
@@ -0,0 +1,113 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:nil -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <iostream>
+
+#include <seastar/core/app-template.hh>
+#include <seastar/core/thread.hh>
+
+#include "common/ceph_argparse.h"
+#include "crimson/common/config_proxy.h"
+#include "crimson/mon/MonClient.h"
+#include "crimson/net/SocketMessenger.h"
+
+#include "osd.h"
+
+using config_t = ceph::common::ConfigProxy;
+
+void usage() {
+  std::cout << "usage: crimson -i <ID>" << std::endl;
+  generic_server_usage();
+}
+
+int main(int argc, char* argv[])
+{
+  std::vector<const char*> args{argv + 1, argv + argc};
+  if (args.empty()) {
+    std::cerr << argv[0] << ": -h or --help for usage" << std::endl;
+    return EXIT_FAILURE;
+  }
+  std::string cluster;
+  std::string conf_file_list;
+  // ceph_argparse_early_args() could _exit(), while local_conf() won't ready
+  // until it's started. so do the boilerplate-settings parsing here.
+  auto init_params = ceph_argparse_early_args(args,
+                                              CEPH_ENTITY_TYPE_CLIENT,
+                                              &cluster,
+                                              &conf_file_list);
+  if (ceph_argparse_need_usage(args)) {
+    usage();
+    return EXIT_SUCCESS;
+  }
+
+  int return_value = 0;
+  // TODO: add heartbeat
+  seastar::app_template app;
+  // talk with osd
+  ceph::net::SocketMessenger cluster_msgr;
+  // talk with mon/mgr
+  ceph::net::SocketMessenger client_msgr;
+  ceph::mon::Client monc{client_msgr};
+  seastar::sharded<OSD> osd;
```

Comment:
Agree with the sharded OSD design, if we want to stick to run-to-completion from crimson-msgr to crimson-OSD, each sharded OSD only needs to handle a subset of connections until dispatched to PGs.

One effect in my mind is that each sharded OSD only needs to manage a subset of sessions now, this requires shared-nothing design for those sessions in the OSD. For the fact that each sharded OSD is only running in one core, it also brings us opportunity to revisit that if we still need locks related to OSD and sessions.

Reply:
I think Jianpeng's meaning is in which phase do OSD_init works. it is process global,  for example mount memstore in our POC case, so I think these kind of works can be done before seastar reactor start, then start OSD shard do other network io job.

Reply:
> I think these kind of works can be done before seastar reactor start, 

how can we read from config without ceph::common::local_conf?

Reply:
why would we do the initialization without reactor?